qoffer-infrastructure/src/main/groovy/life/qbic/portal/qoffer2/web/viewmodel/CreateOfferViewModel.groovy

```diff
@@ -37,32 +41,101 @@ class CreateOfferViewModel {
+    @Bindable
+    double totalPrice = 0
 
     private final PersonResourcesService personService
+    private final ProductsResourcesService productsResourcesService
 
-    CreateOfferViewModel(PersonResourcesService personService) {
+    CreateOfferViewModel(PersonResourcesService personService, ProductsResourcesService productsResourcesService) {
         this.personService = personService
-        this.availableProjectManagers = personService.getProjectManagers()
-        this.foundCustomers = personService.getCustomers()
-        this.personService.customerEvent.register( (List<Customer> customerList) -> {
+        this.productsResourcesService = productsResourcesService
+        fetchPersonData()
+        fetchProductData()
+        subscribeToResources()
+    }
+
+    private void fetchPersonData() {
+        this.availableProjectManagers.clear()
+        this.availableProjectManagers.addAll(personService.getProjectManagers())
+        this.foundCustomers.clear()
+        this.foundCustomers.addAll(personService.getCustomers())
+    }
+
+    private void fetchProductData() {
+        populateProductLists(productsResourcesService.getProducts())
+    }
+
+    private void subscribeToResources() {
+        this.personService.customerEvent.register((List<Customer> customerList) -> {
             this.foundCustomers.clear()
             this.foundCustomers.addAll(customerList)
         })
         this.personService.projectManagerEvent.register((List<ProjectManager> managerList) -> {
             this.availableProjectManagers.clear()
             this.availableProjectManagers.addAll(managerList)
         })
+
+        Subscription<List<Product>> productSubscription = new Subscription<List<Product>>() {
+            @Override
+            void receive(List<Product> products) {
+                populateProductLists(products)
+            }
+        }
+        this.productsResourcesService.productEventEmitter.register(productSubscription)
     }
 
+    private void populateProductLists(List<Product> products) {
+        this.sequencingProducts.clear()
+        this.managementProducts.clear()
+        this.primaryAnalysisProducts.clear()
+        this.secondaryAnalysisProducts.clear()
+        this.storageProducts.clear()
+
+        products.each { product ->
+            ProductItemViewModel productItem = new ProductItemViewModel(0, product)
+
+            if (product instanceof Sequencing) {
+                this.sequencingProducts.add(productItem)
+            } else if (product instanceof ProjectManagement) {
+                this.managementProducts.add(productItem)
+            } else if (product instanceof PrimaryAnalysis) {
+                this.primaryAnalysisProducts.add(productItem)
+            } else if (product instanceof SecondaryAnalysis) {
+                this.secondaryAnalysisProducts.add(productItem)
+            } else if (product instanceof DataStorage) {
+                this.storageProducts.add(productItem)
+            }
```

Comment:
```suggestion
           switch(product) {
               case Sequencing:
                   sequencingProducts.add(productItem)
                   break
               case ProjectManagement:
                   managementProducts.add(productItem)
                   break
               case PrimaryAnalysis:
                   primaryAnalysisProducts.add(productItem)
                   break
               case SecondaryAnalysis:
                   secondaryAnalysisProducts.add(productItem)
                   break
                case DataStorage:
                   storageProducts.add(productItem)
                   break
           }
```

Reply:
more readable

Reply:
Sure this is a nice idea. I did not change that because I only moved existing code as to not introduce new bugs in the process. If this works we could do it as you suggested :+1:

Reply:
cf8d02c